src/openms_gui/source/VISUAL/TVToolDiscovery.cpp

```diff
@@ -0,0 +1,125 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: David Voigt $
+// $Authors: David Voigt $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/VISUAL/TVToolDiscovery.h>
+#include <OpenMS/APPLICATIONS/ToolHandler.h>
+#include <OpenMS/SYSTEM/File.h>
+#include <OpenMS/FORMAT/ParamXMLFile.h>
+
+#include <QProcess>
+#include <QCoreApplication>
+
+#include <iostream>
+
+namespace OpenMS {
+
+    void TVToolDiscovery::loadParams()
+    {
+      // tool params are only loaded once by using a immediately evaluated lambda
+      static bool _ [[maybe_unused]] = [&]() -> bool
+      {
+        // Get a map of all tools
+        const auto& tools = ToolHandler::getTOPPToolList();
+        const auto& utils = ToolHandler::getUtilList();
+        // Launch threads for loading tool/util params.
+        for (auto& [name, description] : tools)
+        {
+          param_futures_[name] = std::async(std::launch::async, getParamFromIni_, name);
+        }
+        for (auto& [name, description] : utils)
+        {
+          param_futures_[name] = std::async(std::launch::async, getParamFromIni_, name);
+        }
+        return true;
+      }();
+    }
+
+    void TVToolDiscovery::waitForParams()
+    {
+      // Make sure that future results are only waited for and inserted in params_ once
+      static bool _ [[maybe_unused]] = [&]() -> bool
+      {
+        // Make sure threads have been launched before waiting
+        loadParams();
+        // Wait for futures to finish
+        for (auto& [name, param_future] : param_futures_)
+        {
+          while (param_future.wait_for(std::chrono::milliseconds(10)) != std::future_status::ready)
+          {
+            // Keep GUI responsive while waiting
+            QCoreApplication::processEvents();
+          }
+          // Make future results available in params_
+          params_.emplace(name, param_future.get());
+        }
+        return true;
+      }();
+    }
+
+    const std::map<std::string, Param>& TVToolDiscovery::getToolParams()
+    {
+      // Make sure threads have been launched and waited for before accessing results
+      loadParams();
+      waitForParams();
+      return params_;
+    }
+
+    Param TVToolDiscovery::getParamFromIni_(const std::string& tool_name)
```

Comment:
this whole function looks very `manual`.
Look for higher order functions, such as `File::getTemporaryFile` or `ExternalProcess::run` to use a lot less code..

Reply:
I don't think `File::getTemporaryFile` is thread-safe. Having the path constructed in a "hacky" way is probably to be preferred over rewriting `File::getTemporaryFile` and all associated functions just for this one use case where it has to be thread-safe. What do you think?
You're right though, it is kind of an ugly function. I'll look into replacing some with preexisting functions.

Reply:
hmm yes doesn't look thread safe to me but this can be easily fixed in that case as far as I see.
add:
std::lock_guard<std::mutex> lk(class_mutex_); 
as first line of every member function of TemporaryFiles_
and a private member std::mutex class_mutex_;

Reply:
I've looked into `ExternalProcess::run` as well. However the problem with too many open files handles I described previously reoccurred. I could fix that by adding an additional bool parameter with a default value (so that no existing calls need to be changed) of false to `ExternalProccess::run` like `read_only` so that
```cpp
ExternalProcess::RETURNSTATE ExternalProcess::run(const QString& exe, const QStringList& args, const QString& working_dir, const bool verbose, String& error_msg, bool read_only)
{
  // ...
  qp_->start(exe, args, read_only? QProcess::NotOpen : QProcess::ReadWrite) // opposed to qp_->start(exe, args)
  // ...
}
```

Would that be a possible workaround?

Reply:
I think this should work. Maybe call the parameter different in our API e.g. noIO and add a comment/documentation that this leads to fewer threads opened by the process?

Reply:
concerning thread-safe `getTemporaryFile`: you should get away with `static std::atomic<int> number = 0;` in `File::getUniqueName` and a thread-local storage of `TemporaryFiles_::filenames_`

Reply:
concerning `ExternalProcess::run`: bool parameters are evil (because you never know what `someFunc(true, true, false)` is actually doing unless you look it up. Better is `someFunc(std::ios::in, MyEnum::truncate, Log::verbose)`, since its much more informative.
So, I'd suggest to just pass the parameter right through and make `QIODevice::OpenModeFlag mode = QIODevice::ReadWrite` the default. This incurs adding a Qt include which is not optimal. You could avoid this by defining your own enum with common modes and translate that to the Qt mode inside the cpp. This way you have the best of both worlds.

Reply:
> concerning thread-safe `getTemporaryFile`: you should get away with `static std::atomic<int> number = 0;` in `File::getUniqueName` and a thread-local storage of `TemporaryFiles_::filenames_`

I think both works. Class level locking is simpler but slower. Though the overhead here is probably never an issue.

Reply:
> concerning thread-safe `getTemporaryFile`: you should get away with `static std::atomic<int> number = 0;` in `File::getUniqueName` and a thread-local storage of `TemporaryFiles_::filenames_`

Can you elaborate how you would go about making `TemporaryFiles_::filenames_` thread-local and how that would help with thread-safety while still maintaining the same functionality?

 As far as I understood it a `thread_local` leads to each thread having it's own copy of a variable as opposed to locking the `filenames_` vector behind a `mutex` where each thread still accesses the same vector. The first approach seems odd since this would mean that each thread creates new temporary files and pushes the filename onto a thread local copy of `filenames_` so that upon destruction not all temporary file will be deleted because all the created filenames are scattered across different copies of `filenames_`.
Also [a `thread_local` data member has also be declared static](https://stackoverflow.com/a/32365713). This will probably only complicate things even more. 

A different and probably much simpler idea is just to generate a unique folder in `/tmp` (or whatever the windows and macOS equivalent is) like `/tmp/OpenMS{CurrentDate}{CurrentTime}` and put all temporary files in there and just delete the whole folder upon destruction. That way it's inherently safe and no mutex is needed because it's not necessary to keep track of all filenames. What do you think of that?